option.go

```diff
@@ -166,6 +166,25 @@ func WithRequireConnectProtocolHeader() HandlerOption {
 	return &requireConnectProtocolHeaderOption{}
 }
 
+// WithHandlerOptional is a function that returns a HandlerOption. It's used
+// to conditionally apply HandlerOption to a Handler based on the Spec.
+// For example, to apply a HandlerOption only to a specific procedure:
+//
+//	connect.WithHandlerOptional(func(spec connect.Spec) connect.HandlerOption {
+//		if spec.Procedure == pingv1connect.PingServicePingProcedure {
+//			return connect.WithReadMaxBytes(1024)
+//		}
+//		return nil
+//	})
+type WithHandlerOptional func(Spec) HandlerOption
```

Comment:
Unfortunately, we've already got a `WithHandlerOptions` - it turns a slice of handler options into a single option (which is useful with Ed's design). I don't mind Josh's version either, though - up to you guys which you prefer.

What do you think of the name `WithConditionalHandlerOptions`? Might be worth shipping `WithConditionalOptions` and `WithConditionalClientOptions` too.

Reply:
I think the single `HandlerOption` function with a single return allows for better configuration. For example if you have a config and want to optionally apply handler options this can be done in a single option:

```go
connect.WithHandlerOption(func(spec connect.Spec) connect.HandlerOption {
  var options []connect.HandlerOption
  if cfg, ok := config[spec.Procedure]; ok {
    if cfg.ReadMaxLimit > 0 {
      options = append(options, connect.WithReadMaxBytes(cfg.ReadMaxBytes)
    }
    // ...
  }
  return connect.WithHandlerOptions(options...)
})
```
To avoid the nil case maybe the signature should be:
 ```go
func WithHandlerOptional(f func(Spec) []HandlerOption) HandlerOption
```
Which helps point out that there can be zero or more options returned.

Reply:
That signature feels a little overly tailored to the above use case (where you want to produce per-method options from some separate config source), and might be a little less intuitive for simpler usages. But if we think that's a likely usage pattern, then it is indeed simpler to configure with that signature.

So having the function return a slice, along with a name like `WithConditionalHandlerOptions`, is a marked improvement IMO. 👍